### ISSUE TITLE
Installation issue: hdf5+fortran %oneapi

### ISSUE THREAD

**Issue by cessenat** [LLNL/spack#22721]
Despite PR #22683, I still fail to compile HDF5+fortran with Intel OneAPI, either 2021.1.1 or 2021.2.0.
This is both on Ubuntu x86_64 18.04 and 20.04 
Basic error remains the same : there is a failing -loopopt=0 from configure.
$ spack install hdf5+fortran~mpi+shared %oneapiAT2021.2
==> Installing zlib-1.2.11-jesz5iyurv6flmlwyt3lolavgui6f6vn
==> No binary for zlib-1.2.11-jesz5iyurv6flmlwyt3lolavgui6f6vn found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/c3/c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1.tar.gz
################################################################################################################ 100.0%
==> No patches needed for zlib
==> zlib: Executing phase: 'install'
==> zlib: Successfully installed zlib-1.2.11-jesz5iyurv6flmlwyt3lolavgui6f6vn
  Fetch: 2.82s.  Build: 2.51s.  Total: 5.33s.
[+] /tmp/spack/opt/spack/linux-ubuntu18.04-zen2/oneapi-2021.2/zlib-1.2.11-jesz5iyurv6flmlwyt3lolavgui6f6vn
==> Installing hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty
==> No binary for hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/7a/7a1a0a54371275ce2dfc5cd093775bb025c365846512961e7e5ceaecb437ef15.tar.gz
################################################################################################################ 100.0%
==> Ran patch() for hdf5
==> hdf5: Executing phase: 'autoreconf'
==> hdf5: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/spack-src/configure' '--prefix=/tmp/spack/opt/spack/linux-ubuntu18.04-zen2/oneapi-2021.2/hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty' '--enable-unsupported' '--enable-symbols=yes' '--with-zlib=/tmp/spack/opt/spack/linux-ubuntu18.04-zen2/oneapi-2021.2/zlib-1.2.11-jesz5iyurv6flmlwyt3lolavgui6f6vn' '--disable-threadsafe' '--disable-cxx' '--disable-hl' '--enable-fortran' '--disable-java' '--without-szlib' '--enable-build-mode=production' '--enable-shared' 'CFLAGS=-fPIC' 'CXXFLAGS=-fPIC' 'FCFLAGS=-fPIC'

2 errors found in build log:
     86    checking what /tmp/spack/lib/spack/env/oneapi/ifx does with modules... module.mod
     87    checking how /tmp/spack/lib/spack/env/oneapi/ifx finds modules... -I
     88    checking if Fortran compiler version compatible with Fortran 2003 HDF... yes
     89    checking how to get verbose linking output from /tmp/spack/lib/spack/env/oneapi/ifx... -v
     90    checking for Fortran libraries of /tmp/spack/lib/spack/env/oneapi/ifx...  **-loopopt=0** -L/opt/intel/oneapi/co
           mpiler/2021.2.0/linux/compiler/lib/intel64_lin -L/opt/intel/oneapi/compiler/2021.2.0/linux/lib -L/usr/lib/g
           cc/x86_64-linux-gnu/7/ -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/ -L/usr/lib/gcc/x86_64-l
           inux-gnu/7/../../../../lib/ -L/lib/x86_64-linux-gnu/ -L/lib/../lib64 -L/lib/../lib/ -L/usr/lib/x86_64-linux
           -gnu/ -L/usr/lib/../lib/ -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../ -L/lib64 -L/lib/ -L/usr/lib -L/usr/lib
           /i386-linux-gnu -lifport -lifcoremt -limf -lsvml -lm -lipgo -lirc -lpthread -lirc_s -ldl
     91    checking for dummy main to link with Fortran libraries... unknown
  >> 92    configure: error: in `/tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/sp
           ack-src':
  >> 93    configure: error: linking to Fortran libraries from C fails
     94    See `config.log' for more details
     94    See `config.log' for more details

See build log for details:
  /tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/spack-build-out.txt

Also, there are weird errors with
- OneAPI 2021.1.1:
$ spack install hdf5ATdevelop+fortran~mpi+shared %oneapiAT2021.1 ^autoconf@2.71
... bla bla bla ...
     9058    libtool: compile:  /home/cessenat/Softs/spack/lib/spack/env/oneapi/ifx -I. -
             I../../src -stand:f03 -free -warn:all -g -O3 -I../../src -I../../fortran/src
              -fPIC -c HDF5.F90 -o HDF5.o >/dev/null 2>&1
     9059      FCLD     libhdf5_fortran.la
     9060    ld: .libs/H5_ff.o: **relocation R_X86_64_PC32 against symbol `h5global_mp_pred
             ef_types_' can not be used when making a shared object; recompile with -fPIC**
  >> 9061    ld: final link failed: bad value
     9062    make[3]: *** [Makefile:987: libhdf5_fortran.la] Error 1
     9063    make[3]: Leaving directory '/tmp/cessenat/spack-stage/spack-stage-hdf5-devel
             op-yxj7hfd2ulqg2sdkboqozeufaaullqfp/spack-src/fortran/src'

- OneAPI 2021.2.0:
$ spack install hdf5ATdevelop+fortran~mpi+shared %oneapiAT2021.2 ^autoconf@2.71
... bla bla bla ...
  >> 9286    tf.F90(150): error #5528: The REFERENCE attribute cannot be used with a pass
             ed length CHARACTER variable on this platform
     9287           INTEGER FUNCTION h5_fixname_c(base_name, base_namelen, fapl, &
     9288    -------------------------------------^
  >> 9289    tf.F90(151): error #5528: The REFERENCE attribute cannot be used with a pass
             ed length CHARACTER variable on this platform
     9290                full_name, full_namelen)
     9291    ------------^
  >> 9292    tf.F90(207): error #5528: The REFERENCE attribute cannot be used with a pass
             ed length CHARACTER variable on this platform
     9293           INTEGER FUNCTION h5_cleanup_c(base_name, base_namelen, fapl)
     9294    -------------------------------------^
     9295    compilation aborted for tf.F90 (code 1)

* **Spack:** 0.16.1-2002-09dcb16a70
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-zen2
* **Concretizer:** original

~/.spack/linux/compilers.yaml
- compiler:
    spec: oneapi@2021.2
    paths:
      cc: /opt/intel/oneapi/compiler/2021.2.0/linux/bin/icx
      cxx: /opt/intel/oneapi/compiler/2021.2.0/linux/bin/icpx
      f77: /opt/intel/oneapi/compiler/2021.2.0/linux/bin/ifx
      fc: /opt/intel/oneapi/compiler/2021.2.0/linux/bin/ifx
    flags: {}
    operating_system: ubuntu18.04
    target: x86_64
    modules: [compiler/2021.2.0]
    environment:
      append_path:
        LD_LIBRARY_PATH: ${LD_LIBRARY_PATH}:/opt/intel/oneapi/compiler/2021.2.0/linux/compiler/lib/intel64_lin
      set:
        MODULEPATH: ${MODULEPATH}:/opt/intel/oneapi/modulefiles
    extra_rpaths:
    - /opt/intel/oneapi/compiler/2021.2.0/linux/compiler/lib/intel64_lin

### Additional information

* [spack-build-out.txt](
[spack-build-out.txt](https://github.com/spack/spack/files/6243465/spack-build-out.txt)
) 
[spack-build-out.txt](https://github.com/spack/spack/files/6243589/spack-build-out.txt)

* [spack-build-env.txt](
[spack-build-env.txt](https://github.com/spack/spack/files/6243466/spack-build-env.txt)
)
[spack-build-env.txt](https://github.com/spack/spack/files/6243591/spack-build-env.txt)

Addon:
 -error for "spack install hdf5ATdevelop+fortran~mpi+shared %oneapiAT2021.1 ^autoconf@2.71":
[hdf5_spack-build-out_oneapi2021-1.txt](https://github.com/spack/spack/files/6243529/hdf5_spack-build-out_oneapi2021-1.txt)
[hdf5_spack-build-env_oneapi2021-1.txt](https://github.com/spack/spack/files/6243544/hdf5_spack-build-env_oneapi2021-1.txt)

 -error for "spack install hdf5ATdevelop+fortran~mpi+shared %oneapiAT2021.2 ^autoconf@2.71":
[hdf5_spack-build-out_oneapi2021-2.txt](https://github.com/spack/spack/files/6243531/hdf5_spack-build-out_oneapi2021-2.txt)
[hdf5_spack-build-env_oneapi2021-2.txt](https://github.com/spack/spack/files/6243546/hdf5_spack-build-env_oneapi2021-2.txt)

@lrknox @tgamblin @frankwillmore

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by cessenat** [comment]
Same  -loopopt=0 for silo.

**Comment by cessenat** [comment]
Same problem with onapi 2021.1.2 internal to spack on another platform :
spack debug report
* **Spack:** 0.16.1-2030-9f90659
* **Python:** 3.6.8
* **Platform:** linux-centos7-skylake
* **Concretizer:** original

**Comment by MBlaschek** [comment]
Same problem with oneapi 2021.2.0, installed with spack:
* **Spack:** 0.16.0-2328-d64ef5fbe6
* **Python:** 3.6.8
* **Platform:** linux-rhel8-skylake_avx512
* **Concretizer:** original

**Comment by MBlaschek** [comment]
OK. I checked the config.log file and it says:
```
configure:7549: /jetfs/spack/opt/spack/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-mpi-2021.2.0-haqpxflbk23spgt
k7x74tgecneqnya57/mpi/2021.2.0/bin/mpicc -o conftest  -fPIC     conftest.c   -loopopt=0 -LI_MPI_SUBSTITUTE_INSTALLDIR/lib/r
elease -LI_MPI_SUBSTITUTE_INSTALLDIR/lib -L/jetfs/spack/opt/spack/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-c
ompilers-2021.2.0-6kdzddxrshhliyowfwtuiyzjo4hw7dro/compiler/2021.2.0/linux/compiler/lib/intel64_lin -L/jetfs/spack/opt/spac
k/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-compilers-2021.2.0-6kdzddxrshhliyowfwtuiyzjo4hw7dro/compiler/2021
.2.0/linux/lib -L/jetfs/spack/opt/spack/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-compilers-2021.2.0-6kdzddxr
shhliyowfwtuiyzjo4hw7dro/lib -L/usr/lib/gcc/x86_64-redhat-linux/8/ -L/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64 -
L/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/ -L/lib/../lib64 -L/lib/../lib64/ -L/usr/lib/../lib64 -L/usr/lib/../l
ib64/ -L/jetfs/spack/opt/spack/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-compilers-2021.2.0-6kdzddxrshhliyowf
wtuiyzjo4hw7dro/compiler/2021.2.0/linux/compiler/lib/intel64_lin/ -L/jetfs/spack/opt/spack/linux-rhel8-skylake_avx512/oneap
i-2021.2.0/intel-oneapi-compilers-2021.2.0-6kdzddxrshhliyowfwtuiyzjo4hw7dro/compiler/2021.2.0/linux/lib/ -L/jetfs/spack/opt
/spack/linux-rhel8-skylake_avx512/oneapi-2021.2.0/intel-oneapi-compilers-2021.2.0-6kdzddxrshhliyowfwtuiyzjo4hw7dro/lib/ -L/
usr/lib/gcc/x86_64-redhat-linux/8/../../../ -L/lib64 -L/lib/ -L/usr/lib64 -L/usr/lib -lmpifort -lmpi -ldl -lrt -lpthread -l
ifport -lifcoremt -limf -lsvml -lm -lipgo -lirc -lirc_s >&5
/usr/bin/ld: cannot find -loopopt=0
clang: error: linker command failed with exit code 1 (use -v to see invocation)
configure:7549: $? = 1
```
But I do not know what that means.

**Comment by MBlaschek** [comment]
found this on [intel-community](https://community.intel.com/t5/Intel-Fortran-Compiler/ifx-2021-1-beta04-HPC-Toolkit-build-error-with-loopopt/m-p/1184181), suggesting to update autoconf 

> Autoconf 2.70 was released today (see https://lists.gnu.org/archive/html/autotools-announce/2020-12/msg00001.html for details).  Updating to autoconf 2.70 fixes the “-loopopt=0” issue described in this discussion.

I was using autoconf 2.69....trying it right now.

**Comment by frankwillmore** [comment]
if you re-run autogen.sh in hdf5 with the new autoconf, it will get rid of loopopt. There's also a PR that has the compiler wrapper edit it out: https://github.com/spack/spack/pull/23482 . Available in 0.16.2

**Comment by cessenat** [comment]
Thank you @frankwillmore I no longer have the problem.
Yet there is one more compilation error, both with the preferred version and with the develop one (I did not check more versions):
```
6 errors found in build log:
     11390    tf.F90(404): warning #7416: Fortran 2003 does not allow this intrinsic procedure.   [STORAGE_SIZE]
     11391        H5_SIZEOF_DP = storage_size(a, c_size_t)/storage_size(c_char_'a',c_size_t)
     11392    -------------------^
     11393    tf.F90(404): warning #7416: Fortran 2003 does not allow this intrinsic procedure.   [STORAGE_SIZE]
     11394        H5_SIZEOF_DP = storage_size(a, c_size_t)/storage_size(c_char_'a',c_size_t)
     11395    ---------------------------------------------^
  >> 11396    tf.F90(150): error #5528: The REFERENCE attribute cannot be used with a passed length CHARACTER variable on this platform
     11397           INTEGER FUNCTION h5_fixname_c(base_name, base_namelen, fapl, &
     11398    -------------------------------------^
  >> 11399    tf.F90(151): error #5528: The REFERENCE attribute cannot be used with a passed length CHARACTER variable on this platform
     11400                full_name, full_namelen)
     11401    ------------^
  >> 11402    tf.F90(207): error #5528: The REFERENCE attribute cannot be used with a passed length CHARACTER variable on this platform
     11403           INTEGER FUNCTION h5_cleanup_c(base_name, base_namelen, fapl)
     11404    -------------------------------------^
     11405    compilation aborted for tf.F90 (code 1)
     11406    Makefile:962: recipe for target 'tf.lo' failed
  >> 11407    make[2]: *** [tf.lo] Error 1
     11408    make[2]: Leaving directory '/tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/spack-src/fortran
              /test'
     11409    Makefile:829: recipe for target 'all-recursive' failed
  >> 11410    make[1]: *** [all-recursive] Error 1
     11411    make[1]: Leaving directory '/tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/spack-src/fortran
              '
     11412    Makefile:669: recipe for target 'all-recursive' failed
  >> 11413    make: *** [all-recursive] Error 1

See build log for details:
  /tmp/cessenat/spack-stage/spack-stage-hdf5-1.10.7-rypct3qztneeyaxk3ihv2vf7ttxc5hty/spack-build-out.txt

```
But this is not related to the initial issue.

**Comment by frankwillmore** [comment]
I believe that's a oneapi compiler issue and IIRC its fixed in an upcoming compiler release. If you use ifort instead of ifx, it will build. Also, FYI, there's an HDF5 issue https://github.com/spack/spack/pull/23482 which you might come across. This is also fixed and should appear in 1.10.8. 

Good luck!

**Comment by cessenat** [comment]
Great ! Many thanks.
